integrations/datadog/client.py

```diff
@@ -233,6 +296,117 @@ async def get_slo_history(
+        """
+        logger.debug(f"Creating query with values: {metric_query} | {variable_values}")
+        for variable_name, value in variable_values.items():
+            metric_query = metric_query.replace(
+                f"${variable_name}", f"{variable_name}:{value}"
+            )
+        return metric_query
+
+    def extract_metric_name(self, query_string):
+        """Extracts the Datadog metric name from a query string.
+
+        Args:
+            query_string: The Datadog query string.
+
+        Returns:
+            The metric name, or None if no metric name is found.
+
+        Examples:
+            - "sum:container.memory.usage" -> "container.memory.usage"
+            - "avg:system.cpu.used{service:my-service}" -> "system.cpu.used"
+        """
+        match = re.search(r"(?:\w+:)?(\w+\.\w+\.\w+)", query_string)
+        if match:
+            return match.group(1)
+        else:
+            return None
+
+    async def get_single_service(self, service_id: str) -> dict[str, Any] | None:
+        if not service_id:
+            return None
+        url = f"{self.api_url}/api/v2/services/definitions/{service_id}"
+        return await self._send_api_request(url)
+
+    async def get_metric_metadata(self, metric: str) -> dict[str, Any] | None:
+        url = f"{self.api_url}/api/v1/metrics/{metric}"
+        return await self._send_api_request(url)
+
+    async def get_metrics(
+        self, query: str, env: str = "prod", service: str = "*"
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        """
+        Fetches metrics for specified services and environment.
+
+        Args:
+            query (str): The Datadog query string to execute.
+            env (str): The environment to filter by (default: "prod").
+            service (str): The service ID to filter by, or "*" to fetch metrics for all services.
+
+        Yields:
+            AsyncGenerator[list[dict[str, Any]], None]: A list of metrics for each service.
+        """
+        url = f"{self.api_url}/api/v1/metrics"
+
+        metrics = []
+
+        services = []
+        if service == "*":
+            # get all services
+            async for service_list in self.get_services():
+                services.extend(service_list)
+        else:
+            # get service by id
+            result = await self.get_single_service(service)
+            if not result.get("data"):
+                logger.warning(f"Service with id {service} not found")
+                return
+
+            services = [result.get("data")]
+
+        logger.info(f"Fetching metrics for {len(services)} services")
+
+        for service in services:
+            service_id = service["attributes"]["schema"]["dd-service"]
+
+            params = {"env": env, "service": service_id}
+            query_with_values = self._create_query_with_values(
+                f"{query}{{service:{service_id},env:{env}}}", params
+            )
```

Comment:
I think user will want to have a different entity for a service when its in `prod` and when its in `dev`, I would also make sure I list all environments and iterate over them

Reply:
i have done this